automates/program_analysis/CAST2GrFN/visitors/cast_to_agraph_visitor.py

```diff
@@ -0,0 +1,449 @@
+        and returns an AGraph of the graph G as a result.
+        """
+        self.visit_list(self.cast.nodes)
+        A = nx.nx_agraph.to_agraph(self.G)
+        A.graph_attr.update(
+            {
+                "dpi": 227, 
+                "fontsize": 20, 
+                "fontname": "Menlo", 
+                "rankdir": "TB"
+            }
+        )
+        A.node_attr.update(
+            {
+                "shape": "rectangle",
+                "color": "#650021",
+                "style": "rounded",    
+                "fontname": "Menlo"
+            }
+        )
+        return A        
+
+    def to_pdf(self,filename):
+        """Creates a pdf of the generated agraph"""
+        A = self.to_agraph()
+        for node in A.iternodes():
+            node.attr["fontcolor"] = "black"
+            node.attr["style"] = "rounded"
+        A.edge_attr.update({"color": "#650021", "arrowsize": 0.5})
+        A.draw(filename+".pdf",prog="dot")
+
+    @singledispatchmethod
+    def visit(self, node: AstNode):
+        """Generic visitor for unimplemented/incorrect nodes"""
+        print("Not implemented: ",node)
+        print("Type not implemented: ",type(node))
+        return None
+
+    @visit.register
+    def _(self, node: Assignment):
+        """Visits Assignment nodes, the node's UID is returned
+        so it can be used to connect nodes in the digraph"""
+        left = self.visit(node.left)
+        right = self.visit(node.right)
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid,label='Assignment')
+        self.G.add_edge(node_uid,left)
+        self.G.add_edge(node_uid,right)
+        return node_uid
+
+    @visit.register
+    def _(self, node: Attribute):
+        """Visits Attribute nodes, the node's UID is returned
+        so it can be used to connect nodes in the digraph"""
+        value = self.visit(node.value)
+        attr = self.visit(node.attr)
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid,label="Attribute")
+        self.G.add_edge(node_uid,value)
+        self.G.add_edge(node_uid,attr)
+        
+        return node_uid
+
+    @visit.register
+    def _(self, node: BinaryOp):
+        """Visits BinaryOp nodes, the node's UID is returned
+        so it can be used to connect nodes in the digraph"""
+        left = self.visit(node.left)
+        right = self.visit(node.right)
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid, label=node.op)
+        self.G.add_edge(node_uid, left)
+        self.G.add_edge(node_uid, right)
+
+        return node_uid
+
+    @visit.register
+    def _(self, node: Call):
+        """Visits Call (function call) nodes. We check to see
+        if we have arguments to the node and act accordingly. 
+        Appending all the arguments of the function to this node,
+        if we have any. The node's UID is returned."""
+        func = self.visit(node.func)
+        args = []
+        if(node.arguments != []):
```

Comment:
@titomeister this is not the safest way to do this check. If a non-empty non-list type were ever assigned to `node.arguments` this check would still fail. A better version of the check would be to see if `node.arguments` is empty by doing the following:
```python
if len(node.arguments) > 0:
    ...
```

Please convert this check to use the version above and all other similar checks found in this file. Also, parens are not required around the conditional portion of an if-statement in Python and they are considered _non-Pythonic_ so I would appreciate it if they were removed 🙂.

Reply:
Done! I changed them all to reflect this Pythonic version. Thank you for the suggestion. I'm not sure where I picked up the habit of using parentheses in my if statements: school or from programming in C...